ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/engine/EngineForkchoiceUpdated.java

```diff
@@ -60,101 +62,108 @@ public String getName() {
-                        payloadAttributes.getSuggestedFeeRecipient()));
-
-        payloadId.ifPresent(
-            pid ->
-                debugLambda(
-                    LOG,
-                    "returning identifier {} for requested payload {}",
-                    pid::toHexString,
-                    () ->
-                        optionalPayloadAttributes.map(
-                            EnginePayloadAttributesParameter::serialize)));
-
-        return new JsonRpcSuccessResponse(
-            requestContext.getRequest().getId(),
-            new EngineUpdateForkchoiceResult(
-                VALID,
-                result.getNewHead().map(BlockHeader::getHash).orElse(null),
-                payloadId.orElse(null),
-                Optional.empty()));
-      } else if (result.isFailed()) {
-        final Optional<Hash> latestValid = result.getLatestValid();
-        return new JsonRpcSuccessResponse(
-            requestContext.getRequest().getId(),
-            new EngineUpdateForkchoiceResult(
-                INVALID,
-                latestValid.isPresent() ? latestValid.get() : null,
-                null,
-                result.getErrorMessage()));
-      }
+    if (result.isFailed()) {
+      final Optional<Hash> latestValid = result.getLatestValid();
+      return new JsonRpcSuccessResponse(
+          requestId,
+          new EngineUpdateForkchoiceResult(
+              INVALID,
+              latestValid.isPresent() ? latestValid.get() : null,
+              null,
+              result.getErrorMessage()));
     }
 
-    Optional.ofNullable(forkChoice.getHeadBlockHash())
-        .filter(hash -> !hash.equals(Hash.ZERO))
-        .ifPresent(mergeCoordinator::getOrSyncHeaderByHash);
+    // begin preparing a block if we have a non-empty payload attributes param
+    Optional<PayloadIdentifier> payloadId =
+        optionalPayloadAttributes.map(
+            payloadAttributes ->
+                mergeCoordinator.preparePayload(
+                    newHead.get(),
+                    payloadAttributes.getTimestamp(),
+                    payloadAttributes.getPrevRandao(),
+                    payloadAttributes.getSuggestedFeeRecipient()));
+
+    payloadId.ifPresent(
+        pid ->
+            debugLambda(
+                LOG,
+                "returning identifier {} for requested payload {}",
+                pid::toHexString,
+                () -> optionalPayloadAttributes.map(EnginePayloadAttributesParameter::serialize)));
+
+    return new JsonRpcSuccessResponse(
+        requestId,
+        new EngineUpdateForkchoiceResult(
+            VALID,
+            result.getNewHead().map(BlockHeader::getHash).orElse(null),
+            payloadId.orElse(null),
+            Optional.empty()));
+  }
+
+  private boolean isPartOfCanonicalChain(final Hash blockHash) {
+    final Optional<BlockHeader> maybeBlockHeader =
+        protocolContext.getBlockchain().getBlockHeader(blockHash);
```

Comment:
the presence of a header in the blockchain isn't really an indicator it is part of the canonical chain.   It could be a fork for example.   If we want to be sure it is part of the canonical chain we should use `isDescendantOf(finalized, safeBlockHash)`

Reply:
The spec is actually transitive about this.  It says that safeBlockHash should be an ancestor of headBlockHash.  But really it needs to be somehwere between the inclusive range of [finalized .. head]